app/vmalert/manager.go

```diff
@@ -97,19 +98,36 @@ func (m *manager) update(ctx context.Context, path []string, validateTpl, valida
 	for _, og := range m.groups {
 		ng, ok := groupsRegistry[og.ID()]
 		if !ok {
+			startAt := time.Now()
 			// old group is not present in new list
 			// and must be stopped and deleted
 			og.close()
 			delete(m.groups, og.ID())
+			logger.Infof("group %q delete success, took %v seconds", og.Name, time.Since(startAt).Seconds())
 			og = nil
 			continue
 		}
-		og.updateCh <- ng
+		go func(og *Group, ng *Group) {
+			startAt := time.Now()
+			select {
+			case <-ctx.Done():
+				logger.Infof("group %q: context cancelled", og.Name)
+				return
+			case <-og.doneCh:
+				logger.Infof("group %q: received stop signal", og.Name)
+				return
+			case og.updateCh <- ng:
+				logger.Infof("group %q reload success, took %v seconds", og.Name, time.Since(startAt).Seconds())
```

Comment:
it's not actually a success, it is just shows that something consumed from the channel. The real reload may be still in-progress or even fail with error.
With running reload in separate goroutine it is hard to tell when update is done. It also means that two identical groups could run evaluation simultaneously updating the same metrics and sending same rule results to remote storage. Moreover, after call `reload` user may notice success reload logs for 9/10 groups and miss that the least one was reloaded after 30s only. 
I'd imaging some way to interrupt group evaluation process on new update. Not sure how to do this right now, may be some additional ctx that may be issued for every evaluation and cancelled on demand. Not sure. What do you think?

Reply:
> it's not actually a success, it is just shows that something consumed from the channel. The real reload may be still in-progress or even fail with error.

I'll try to move the log into update function

> With running reload in separate goroutine it is hard to tell when update is done. Moreover, after call reload user may notice success reload logs for 9/10 groups and miss that the least one was reloaded after 30s only.

Yes, that 's why I removed the "file reload success" log. About the update progress, how about adding a new metrics to show the progress of the reloading.

Reply:
> It also means that two identical groups could run evaluation simultaneously updating the same metrics and sending same rule results to remote storage. 

I don't understand. 
According to the code, the actual reload and the evaluation is synchronous, when updating the group, evaluation is temporary blocked until the update is done(or fail). If multi group reload signal is pending, it will run update until all the signal are consumed, thus, it update the group to the newest version.

Reply:
Yeah, sorry. I was wrong.